### ISSUE TITLE
Improve/document API for binding a volume to a container

### ISSUE THREAD

**Issue by LyndonArmitage** [spotify/docker-client#228]
Given that there are `volumesFrom()` methods on the `HostConfig.Builder`, `volumes()` methods on the `ContainerConfig.Builder` and `binds()` methods on the `HostConfig.Builder`, all of which missing any JavaSoc, I spent some time trying to figure out the correct one to use in order to mount a logs directory for a tomcat container.

An example in the readme and some JavaDoc on these methods would go a long way to alleviate any confusion to others.

Also it might be worth improving the binds API by creating a simple Java object for a mounts settings, similar to a [Pair](http://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/tuple/Pair.html) object or altering the binds method to take 2 parameters. So instead of doing something like this:

```java
HostConfig hostConfig = HostConfig.builder()
		.portBindings(getPortBindings(usedPort))
		.binds(logsDir.getAbsolutePath() + ":/usr/local/tomcat/logs") // binds docker volumes
		.build();
```
You'd instead do this:

```java
HostConfig hostConfig = HostConfig.builder()
		.portBindings(getPortBindings(usedPort))
		.binds(logsDir.getAbsolutePath(), "/usr/local/tomcat/logs") // binds docker volumes
		.build();
```

or this:

```java
HostConfig hostConfig = HostConfig.builder()
		.portBindings(getPortBindings(usedPort))
		.binds(new MountPair(logsDir.getAbsolutePath(), "/usr/local/tomcat/logs")) // binds docker volumes
		.build();
````

**Comment by ymruan** [comment]
totally agree.

I firstly went wrong with following. 
HostConfig hostConfig = HostConfig.builder()
        .portBindings(getPortBindings(usedPort))
        .binds(logsDir.getAbsolutePath(), "/usr/local/tomcat/logs") // binds docker volumes
        .build();

**Comment by grexe** [comment]
yes the lack of any documentation (besides very sparse Javadoc) also led me here (what about a project wiki?).
I'd propose to stay lightweight and follow the builder pattern like:

    bindVolume(from("/path/on/host/").to("/path/in/container")
    .bindVolune(from("/another/host/path").to("/another/container/path")

**Comment by grexe** [comment]
Just noticed that in v3.2.1, binds() is just a List that allows easy adding of volume mounts (verbatim, the syntax could be hidden by the API, but it's easy enough), so that's fine with me...

**Comment by johnflavin** [comment]
I'm working on this.

**Comment by johnflavin** [comment]
This is harder than I would have expected (in addition to the fact that writing documentation is always hard). 

1. This issue was created because there are several ways to create containers with volumes in `docker-client`, and it's confusing. The several ways also exist in `docker`, and it isn't any less confusing to other people who directly use the docker remote API. There are lots of issues on [docker/docker](https://github.com/docker/docker) with people asking questions about or reporting inconsistencies with containers and volumes in the API. It's hard to find anyone laying out a set of best practices.
2. Several new features related to volumes have been introduced to the remote API in the last few versions. So when researching how the volumes should be mounted, I also have to sort out what features are and aren't supported in `docker-client` and try to add those that aren't. Then document the new features.
3. The remote API documentation doesn't even accurately reflect how to use the remote API to create containers with volumes. I've filed an issue with docker about that: [#21335](https://github.com/docker/docker/issues/21335). But it makes it difficult to know what code to write when I can't just read what the API does in the docs, I have to test everything myself to be sure. (With different API versions, because remember that it's changing all the time.)

So, yeah. This is a thing. :neutral_face:

**Comment by LyndonArmitage** [comment]
Fully expected that it wouldn't be easy! 

The best course of action is to probably continue what you're doing and also try and come up with a few equivalent examples in docker client to the actual API.

It might feel like a bit wrong but also linking the docs up to the relevant parts of the official API docs might also help in some small way; at any rate it would highlight the myriad of ways you can accomplish things so people like me will see it and realise the difficulty in documentation.

Thanks for the effort @johnflavin and I hope they get back to you on that issue.

**Comment by johnflavin** [comment]
I think this can be closed with #378 being merged.

**Comment by mattnworb** [comment]
I agree, thanks again @johnflavin
